### ISSUE TITLE
Take a look at Google's Insights project (https://deps.dev/)

### ISSUE THREAD

**Issue by sschuberth** [oss-review-toolkit/ort#4112]
Google just announced the [Insights](https://opensource.googleblog.com/2021/06/introducing-open-source-insights-project.html) project. In large parts it has similarities with ORT's analyzer and advisor components, and pages like [this one](https://deps.dev/maven/com.vdurmont%3Asemver4j) have similarities with the web app report.

I'll reach out to the Insights team to learn about how they are doing things.

**Comment by sschuberth** [comment]
Their FAQ is a very interesting read:

"[How is Insights implemented?](https://deps.dev/faq#how-is-insights-implemented)

Most of the back end is written in Go and *reimplements the dependency analysis algorithm* (version resolution) of each packaging system. The code has been tested for agreement with the native tooling, but is faster because it is decoupled from package installation. It constructs the graph without actually building and installing the software."

**Comment by tsteenbe** [comment]
Sofar insights projects only npm, Maven, Go modules, and Cargo for which static analysis should work really well (apart from maybe Go). Would be interested to learn how they did Go support and how they are planning to do PyPI.

**Comment by sschuberth** [comment]
I'll share what can be learned from their response once I get it.

**Comment by sschuberth** [comment]
Insight's data is now also [available via BigQuery](https://cloud.google.com/blog/products/identity-security/announcing-open-source-insights-data-in-bigquery). Maybe this data could be used by the Analyzer to speed up the lookup of package metadata.

**Comment by sschuberth** [comment]
> I'll share what can be learned from their response once I get it.

BTW, I never heard back from them, so I just reached out again.

**Comment by sschuberth** [comment]
> You can't get the exact picture of the dependency tree unless you use the native package managers resolution mechanism.

I absolutely agree, and that's why ORT's analyzer has always aimed to use the native package manager's resolution mechanism.

However, it was brought to my attention that several ORT users are not necessarily interested in the *exact* picture, but something "close enough" (that can be determined much faster) is sufficient. From a license compliance aspect, that might even be true, as licenses usually do not change between "similar" package versions. But from a security compliance aspect, that might be a problem, as also patch level package versions may introduce severe security risks.

**Comment by sschuberth** [comment]
Their new API now also [supports PURLs](https://blog.deps.dev/api-v3/).

**Comment by sschuberth** [comment]
Also @jamietanna's [Dependency Management Data (DMD)](https://dmd.tanna.dev/) seems to be [using deps.dev](https://github.com/google/deps.dev/issues/10#issuecomment-1988289046), which might be an interesting integration point.

**Comment by jamietanna** [comment]
Yes, you should be able to use DMD alongside ORT (https://dmd.tanna.dev/cookbooks/getting-started-ort/) in the meantime to get this data!

**Comment by heliocastro** [comment]
@jamietanna Just to mention, the docs are outdated, we moved from ort-extended to ort a while ago.
We have this two images:
**ghcr.io/oss-review-toolkit/ort-minimal**  - The (kind of) minimal image with python, rust, ruby and node
**ghcr.io/oss-review-toolkit/ort** - The one that have everything and is the direct replacement for ort-extended

**Comment by jamietanna** [comment]
> @jamietanna Just to mention, the docs are outdated, we moved from ort-extended to ort a while ago. We have this two images: **ghcr.io/oss-review-toolkit/ort-minimal** - The (kind of) minimal image with python, rust, ruby and node **ghcr.io/oss-review-toolkit/ort** - The one that have everything and is the direct replacement for ort-extended

Thanks! I'm on holiday for the next week, but can look to update when I'm home. Or I'd welcome a Merge Request on GitLab if you were happy to update for me ☺

**Comment by sschuberth** [comment]
Using the [deps.dev API](https://docs.deps.dev/api/) could also help to match user expectation like [this](https://github.com/oss-review-toolkit/ort/issues/9993). NPM itself shows a mix of declared and detected licenses on a package's landing page. Currently, ORT needs to run *both* analyzer and scanner to match this information. However, deps.dev returns the detected licenses also via it's API, [it seems](https://deps.dev/npm/JSONStream/0.10.0). So making use of this could save some users from running the scanner at all.

**Comment by sschuberth** [comment]
So, turns out that deps.dev's [native protocol is gRPC](https://github.com/google/deps.dev/issues/5#issuecomment-1960680714) with protobuf, and the REST API implementation is just generated from it. That's also why there is not primary OpenAPI definition for deps.dev. So we should probably use https://github.com/grpc/grpc-kotlin with https://github.com/google/deps.dev/blob/main/api/v3/api.proto to generate client-side Kotlin code for accessing deps.dev.

**Comment by sschuberth** [comment]
> So we should probably use https://github.com/grpc/grpc-kotlin with https://github.com/google/deps.dev/blob/main/api/v3/api.proto to generate client-side Kotlin code for accessing deps.dev.

Or use https://github.com/square/wire. *Edit:* Or https://github.com/hashgraph/pbj.

**Comment by sschuberth** [comment]
> Or use https://github.com/square/wire.

For the record, I wet for Wire. It's awesome.
